packages/flutter/lib/src/gestures/drag.dart

```diff
@@ -14,51 +14,95 @@ enum _DragState {
   accepted,
 }
 
-/// Signature for when a pointer has contacted the screen and might begin to move.
-typedef void GestureDragDownCallback(Point globalPosition);
-
-/// Signature for when a pointer has contacted the screen and has begun to move.
-typedef void GestureDragStartCallback(Point globalPosition);
-
-/// Signature for when a pointer that is in contact with the screen and moving
-/// in a direction (e.g., vertically or horizontally) has moved in that
-/// direction.
-typedef void GestureDragUpdateCallback(double delta);
-
-/// Signature for when a pointer that was previously in contact with the screen
-/// and moving in a direction (e.g., vertically or horizontally) is no longer in
-/// contact with the screen and was moving at a specific velocity when it
-/// stopped contacting the screen.
-typedef void GestureDragEndCallback(Velocity velocity);
+/// Details for [GestureDragDownCallback].
+class DragDownDetails {
+  /// Creates details for a [GestureDragDownCallback].
+  ///
+  /// The [globalPosition] argument must not be null.
+  DragDownDetails({ this.globalPosition: Point.origin }) {
+    assert(globalPosition != null);
+  }
 
-/// Signature for when the pointer that previously triggered a
-/// [GestureDragDownCallback] did not complete.
-typedef void GestureDragCancelCallback();
+  /// The global position at which the pointer contacted the screen.
+  final Point globalPosition;
+}
 
 /// Signature for when a pointer has contacted the screen and might begin to move.
-typedef void GesturePanDownCallback(Point globalPosition);
+typedef void GestureDragDownCallback(DragDownDetails details);
+
+/// Details for [GestureDragStartCallback].
+class DragStartDetails {
+  /// Creates details for a [GestureDragStartCallback].
+  ///
+  /// The [globalPosition] argument must not be null.
+  DragStartDetails({ this.globalPosition: Point.origin }) {
+    assert(globalPosition != null);
+  }
+
+  /// The global position at which the pointer contacted the screen.
+  final Point globalPosition;
+}
 
 /// Signature for when a pointer has contacted the screen and has begun to move.
-typedef void GesturePanStartCallback(Point globalPosition);
+typedef void GestureDragStartCallback(DragStartDetails details);
+
+/// Details for [GestureDragUpdateCallback].
+class DragUpdateDetails {
+  /// Creates details for a [DragUpdateDetails].
+  ///
+  /// The [delta] and [primaryDelta] arguments must not be null. If
+  /// [primaryDelta] is non-zero, then its value must match one of the
+  /// coordinates of [delta] and the other coordinate must be zero.
```

Comment:
it seems to me it'd be cleaner if primaryDelta was null in the case where we're not trying to match either delta.dx or delta.dy.